Localizes nuget.exe with default, embedded resource assembly lookup

dominoFire opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Client.Engineering/issues/1771

Regression? Last working version: Somewhere in nuget 3.x

## Description

- nuget.exe now has localized strings from localization team 
- Localized resources are now embedded into nuget.exe
- A custom resource lookup is implemented in assembly lookup event: CurrentDomain_AssemblyResolve
- If a string is not found, the current default resource lookup mechanism (a string prefixed with three-letter language name) is used as fallback
- Removed duplicated translated strings in NuGetResources.resx, NuGetCommand.resx; those duplicated strings were removed using a custom program to identify those strings. More info at: https://github.com/NuGet/Entropy/pull/90

**NOTE**: To ease reviewing process, it is encouraged to skip changes in .Designer.cs and .resx files; instead, review the program to identify and remove duplicated strings, linked in Entropy PR.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added: Added unit tests 
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [x] Documentation PR or issue filled: Maybe, we will require a doc issue documenting the resource lookup for nuget.exe and how this affects nuget.exe extensions. It's expected to fallback into current method.
  - **OR**
  - [ ] N/A


### Validation

#### Use Case 1: `nuget.exe help` after this PR on Spanish language

Ran nuget.exe in Windows 11 with Spanish language:

![image](https://user-images.githubusercontent.com/1192347/187483167-3e32dbbd-fdba-468c-8640-096ef861b491.png)


### Use case 2: `nuget.exe help` after this PR on Chinese language

NuGet.exe on Chinese (Traditional) has more localized strings. Console is using code page 65001, changed with command: `chcp 65001`

Before | After
-----|------
![nuget-CHT-before](https://user-images.githubusercontent.com/1192347/187585781-af1b20a3-ff06-4ac5-af61-6da54541e13c.png) | ![nuget-CHT-after](https://user-images.githubusercontent.com/1192347/187586001-2c71d96f-bb15-473b-bcfa-20da268c0e92.png)


#### Use case 3: `nuget.exe restore` on a previously restored project

There are new strings localized. NuGet.Commands string resoures are not not localized in both sides

Before| After
------|------
![nuget restore cached before](https://user-images.githubusercontent.com/1192347/190039535-e25f5a65-159b-4b1a-887e-29f7d5a65bbd.png) | ![nuget restore cached after](https://user-images.githubusercontent.com/1192347/190039691-23cbeb2a-d2f6-49fd-a237-44ad6abf41f9.png)

Commit:
Initial approach
parent 95be29c125f550f1f3e8b9ec40fc947d118af51d
author Fernando Aguilar <feaguila@microsoft.com> 1659511997 -0500
committer Fernando Aguilar <feaguila@microsoft.com> 1662080451 -0700

Use default resource manager for string resources in nuget.exe

***NO_CI*** Remove whitespace

Fixed bug on unreachable strings

Commit:
Tests: More tests

Tests: Added tests

Tests: modified tests

Fixed tests

Fixed more tests

Commit:
Adding NuGet.CommandLine resources to merged nuget.exe

Commit:
Adding more loc lookup code

Commit:
Pipeline mods

Trying to fix loc CI pipeline

Removed unused build variable

Commit:
Used new variable syntax for pipeline variables

Commit:
Hardcoding localized assembly count

Commit:
Included all localized assemblies in nuget.exe

Commit:
Only merging Nuget.exe and NuGet.Commands.dll localized resources

Commit:
Getting resource assemblies from embedded resources

Commit:
Added license to source file

Commit:
Min-changeset

Commit:
Second build to include embedded localized resources

jeffkl commented:
Team Review: Show the before/after for file size and a screenshot showing that strings are now localized

nkolev92 commented on eng/pipelines/templates/Build_and_UnitTest.yml:
```diff
@@ -113,7 +113,7 @@ steps:
     solution: "src\\NuGet.Clients\\NuGet.CommandLine\\NuGet.CommandLine.csproj"
     configuration: "$(BuildConfiguration)"
-    msbuildArguments: "/t:ILMergeNuGetExe /p:ExpectedLocalizedArtifactCount=$(LocalizedLanguageCount) /bl:$(Build.StagingDirectory)\\binlog\\04.ILMergeNuGetExe.binlog"
+    msbuildArguments: "/t:Build,ILMergeNuGetExe /bl:$(Build.StagingDirectory)\\binlog\\04.ILMergeNuGetExe.binlog"
```
Wouldn't this build again?

dominoFire replied:
Yes, I need to build first to get binaries to be localized

Later, when satellite assemblies are built, I need to embed those satellite resource assemblies into nuget.exe; This is why I double-build nuget.exe

nkolev92 replied:
I'm still missing why you need to double build nuget.exe. 

Once the satellite assemblies are build, you should only ilmerge correct?

dominoFire replied:
> Once the satellite assemblies are build, you should only ilmerge correct?

No. Resource assemblies are now included in nuget.exe as Embedded Resources.

- First build creates the nuget.exe binary
- Second build creates the nuget.exe binary with the satellite assemblies embedded, not ilMerged

![image](https://user-images.githubusercontent.com/1192347/187570942-b005f79e-b2bd-4f79-8e1d-38c6d15c09a3.png)


The final ILMerge is for including code dependent assemblies into nuget.exe.

nkolev92 replied:
Thanks for that explanation. 

So you are basically embedding so that in the assembly resolve event you can look up a dependent assembly as a resource, correct?

dominoFire replied:
Correct.

dominoFire commented:
Update: I found a bug for portuguese language, please hold on review.

Commit:
Fixed case for cultures with dashes

Commit:
Tests: Added test

Tests: added tests for internal method

Tests: Skip unsupported locales on mono

Tests: Added one more test to windows-only locale test

dominoFire commented:
@NuGet/nuget-client Ready for another review.

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/Program.cs:
```diff
@@ -227,16 +226,55 @@ private void Initialize(CoreV2.NuGet.IFileSystem fileSystem, IConsole console)
             }
+            // .NET Framework 4.x now triggers AssemblyResolve event for resource assemblies
+            // Catch this event for nuget.exe (NuGet.CommandLine) and NuGet.Command resource assemblies only
+            if (name.Name == "NuGet.resources" || name.Name == "NuGet.Commands.resources")
```
Why do we need to special case NuGet.Commands.resources?

I imagine NuGet.resources is the NuGet.Commandline.resources one?

dominoFire replied:
I just followed a comment from that was part of dotnet list|add command implementation. Maybe @rrelyea has some more context here.

zivkan replied:
Why aren't other assemblies like NuGet.Protocol, NuGet.Configuration, NuGet.Common, etc included as well?

dominoFire replied:
Because if we include all NuGet assemblies, this would result in a larger nuget.exe binary, 13 MB approximately.

Also, I think some core localization assemblies are shipped in NuGet.Localization package. Customers can use that package if they want.

zivkan replied:
When a customer's machine uses a system language other than English, how can a customer using nuget.exe use NuGet.Localization to "fix" messages that are being output in English?

dominoFire replied:
Customers just need to unzip the package, and place nuget.exe alongside of lib\netstandard2.0 folder contents.

nkolev92 replied:
> Customers just need to unzip the package, and place nuget.exe alongside of lib\netstandard2.0 folder contents.

Given that this affects UX, I feel like that needs discussed in a design spec.

dominoFire replied:
I'm not changing any new nuget.exe UX ; I'm just ILmerging as it was before. Why do we need a new UX spec? Do you want a UX design spec for merging all/some/none non-nuget.exe resource assemblies?

nkolev92 replied:
My point is that if you want to tell customers to use `NuGet.Localization` then that's a UX change. We're adding a feature, but we don't have details on how they can use it.

dominoFire replied:
Happy to set up an appointment.

nkolev92 replied:
Just to clarify my comment. 

If you want to recommend that, then we need a spec. 

What prompted the whole discussion was the fact that it isn't obvious that we're only localizing nuget.commands and nuget.commandline only. 
The current implementation doesn't really localize strings from nuget.commands correctly, am that right?

NuGet.Protocol contains a healthy amount of strings that are aimed at the user and it's something that restore will frequently do. 
We should consider adding nuget.protocol as well imo.

dominoFire replied:
> The current implementation doesn't really localize strings from nuget.commands correctly, am that right?

That's correct.

> We should consider adding nuget.protocol as well imo.

We can do that in another PR :)

dominoFire replied:
> If you want to recommend that, then we need a spec.

I am going to enable support for nuget.exe localization; I just tried this approach of placing the satellite assemblies alongside nuget.exe and it didn't work. So, not recommended.

nkolev92 commented on eng/pipelines/templates/Build_and_UnitTest.yml:
```diff
@@ -113,7 +113,7 @@ steps:
     solution: "src\\NuGet.Clients\\NuGet.CommandLine\\NuGet.CommandLine.csproj"
     configuration: "$(BuildConfiguration)"
-    msbuildArguments: "/t:ILMergeNuGetExe /p:ExpectedLocalizedArtifactCount=$(LocalizedLanguageCount) /bl:$(Build.StagingDirectory)\\binlog\\04.ILMergeNuGetExe.binlog"
+    msbuildArguments: "/t:Build,ILMergeNuGetExe /bl:$(Build.StagingDirectory)\\binlog\\04.ILMergeNuGetExe.binlog"
```
Should we have separate calls for the build and ilmerge target. I know that'll affect perf negatively, but might be easier to reason about whenever we encounter failures.

Curious what the folks that have done CI work think. cc @zivkan

dominoFire replied:
I'm good doing a single or a separate pipeline step. @zivkan What do you think?

zivkan replied:
This is the commit that made ilmerge happen as a separate step in the first place: https://github.com/NuGet/NuGet.Client/commit/a746cd6a01dfd22187375ee9c5a0d9b98676020a

It appears to have been done because of loc issues.  If that's no longer a problem, then doing ilmerge in the normal build would be good (less customization the better).

dominoFire replied:
I think we still need to build passes for

1. Build first binaries with English resources
2. Run localize step and produce all additional localized .resources.dll assemblies
3. Build final nuget.exe with embedded localized assemblies

zivkan replied:
Ok, I just took a closer look at this.

Why is the `Build` target needed in this task, when it already ran on line 97's build task?

dominoFire replied:
Because that's the second build. This is the updated list:

1. Build first binaries with English resources (line 97)
2. Run localize step and produce all additional localized .resources.dll assemblies (line 104)
3. Build final nuget.exe with embedded localized assemblies (line 116)

zivkan replied:
You forgot step 4, ilmerge all build assemblies into a single file.

I had a second look, and I see now that we're using the compiler's "embed resource", not ilmerge, for the satellite assemblies.

Nikolche's original comment is about splitting steps 3 (second build) & 4 (ilmerge) into separate steps, meaning I completely misunderstood the conversation last week.  Personally, I don't mind either way.

nkolev92 reviewed (commented):
Looks good. I added a few clarifying questions so I understand why we're doing what's we're doing.

zivkan commented on src/NuGet.Clients/NuGet.CommandLine/Program.cs:
```diff
@@ -227,16 +226,55 @@ private void Initialize(CoreV2.NuGet.IFileSystem fileSystem, IConsole console)
+        private static Assembly LoadAssemblyFromEmbeddedResources(string resourceName)
+        {
+            Assembly resourceAssembly = null;
+            using (var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(resourceName))
```
You could extract `Assembly.GetExecutingAssembly()` into a static readonly field, so it doesn't need to be "calculated" each time. Also, I have a personal preference of `typeof(Program).Assembly` over `GetExecutingAssembly`.

dominoFire replied:
It sounds good to cache Assembly.GetExecutingAssembly(). I will try and update the PR as needed.

dominoFire replied:
Changed for this class. Other parts may be using `Assembly.GetExecutingAssembly()` but, I will not change those, since those operations made more sophisticated stuff.

Commit:
Replaced hardcoded string with NuGetResources.ArgumentNullOrEmpty

Commit:
Removed unneeded localized nuget.exe string resources

Commit:
Caching executing assembly

Commit:
Tests: Forced English output to fix tests

Commit:
Separating build and ilmerge steps

Commit:
Removing test changes. Assuming CI with english language

dominoFire commented:
@NuGet/nuget-client Ready for another review.

nkolev92 commented:
@dominoFire 
One of the follow ups from the previous discussion was: 

> Fernando will add before/after comparisons to see if NuGet.Commands is localized before this change. Focus on nuget.exe restore examples.


Did you try that?

dominoFire commented:
> @dominoFire 
> One of the follow ups from the previous discussion was: 
> 
> > Fernando will add before/after comparisons to see if NuGet.Commands is localized before this change. Focus on nuget.exe restore examples.
> 
> 
> Did you try that?

Yeah, it's on the description. It's use case number 3. The 'assets file has not changed' is a string from NuGet.Commands

nkolev92 commented:
my bad, I missed that, I read the description.

So NuGet.Commands strings are not getting localized at all then?
Shouldn't the assembly resolver logic have ensure that it'd get localized in that case?

dominoFire commented:
> my bad, I missed that, I read the description.
> 
> So NuGet.Commands strings are not getting localized at all then? Shouldn't the assembly resolver logic have ensure that it'd get localized in that case?

Yes! I created a follow-up PR for that: https://github.com/NuGet/NuGet.Client/pull/4802

I just want to have this PR merged to continue on that.

Think of this PR as the work needed to localize nuget.exe

Thinkf of the second PR as the work to fix NuGet.Commands localization.

nkolev92 commented on eng/pipelines/templates/Build_and_UnitTest.yml:
```diff
@@ -109,11 +109,18 @@ steps:
-    msbuildArguments: "/t:ILMergeNuGetExe /p:ExpectedLocalizedArtifactCount=$(LocalizedLanguageCount) /bl:$(Build.StagingDirectory)\\binlog\\04.ILMergeNuGetExe.binlog"
+    msbuildArguments: "/t:Build /bl:$(Build.StagingDirectory)\\binlog\\04.BuildNuGetExe.binlog"
+
+- task: MSBuild@1
```
If you look at the logs for `Build NuGet.exe Localized`, you will see that we're now ilmerging twice. 
We need to build twice, but no need to ilmerge twice.

dominoFire replied:
Ok, now it builds twice and ILMerge once

04.BuildNuGetExe.binlog

![Build2nd](https://user-images.githubusercontent.com/1192347/191164982-44586257-5dc5-48ea-b69a-01ed0cd84e9b.png)

05.ILMergeNuGetExe.binlog

![image](https://user-images.githubusercontent.com/1192347/191165057-74970eff-b1fa-4216-9528-07039405dcf7.png)

See [reference build](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6711002&view=artifacts&pathAsName=false&type=publishedArtifacts)

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/Program.cs:
```diff
@@ -32,8 +32,8 @@ public class Program
 
-
-        private static readonly string ThisExecutableName = typeof(Program).Assembly.GetName().Name;
+        internal static readonly Assembly NuGetExeAssembly = typeof(Program).Assembly;
```
Given that these changes are not going to make any localization work, should we just delayed that for the future PR?

It seems like the new branch you have mostly undoes these, https://github.com/NuGet/NuGet.Client/compare/dev-dominofire-nugetexe-defaultloc...dev-dominofire-nugetexe-loc-mixed-embedding

dominoFire replied:
I moved all required changes into this PR and left remaining changes for NuGet.Commands in the other PR.

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj:
```diff
@@ -84,6 +84,10 @@
+    <!-- nuget.exe localized assemblies -->
+    <EmbeddedResource Include="$(OutputPath)\**\$(AssemblyName).resources.dll" Exclude="@(MergeExclude)" />
+    <!-- NuGet.Commands localized assemblies -->
+    <EmbeddedResource Include="$(OutputPath)\**\NuGet.Commands.resources.dll" Exclude="@(MergeExclude)" />
```
Does this accidentally merge all frameworks of NuGet.Commands localized assemblies or no?

dominoFire replied:
No, it only merges net472 resource assemblies in NuGet.CommandLine case; because this OutputPath refers to NuGet.CommandLine output paths.

![NuGet.CommandLine OutputPath property](https://user-images.githubusercontent.com/1192347/190833747-5ca159cc-fa22-4461-a2aa-523883f8775e.png)

nkolev92 replied:
Resovling.

Commit:
Feedback

- Building twice, ilmerging once
- Removed code not to be used in second PR

dominoFire commented:
Just found out that this breaks nuget.mssigning.exe localization. Investigating.
